### DISCUSSION TITLE
How to use a component interface?

### DISCUSSION THREAD

**Discussion by danj3** [tessi/wasmex#684] [Q&A]
This may be related to #675. I believe this is within the realm @superchris and @tessi are currently working.

I am trying to call a function that is part of an interface that is exported. I'm not sure if I just have the wrong name, or what the name should be. I tried using the WIT parser (output below), and it only shows the package export.

This might be the name to resolve or just the stage of development of wasmex, or a bug. But maybe there is a simple answer (before I waste more time)?

I have a very minimal set up, trying to call a JS component function that is described by an interface. A package export works. But since I can't figure this out maybe it will help someone. I appreciate and guidance.

wasmex is looking really good and thank you for the all the work!

Here is the WIT:

```
package root:component;

world root {
  import wasi:io/poll@0.2.2;

  export run: func() -> string;
  export example:ex2/one;
}
package wasi:io@0.2.2 {
  interface poll {
    resource pollable;

    poll: func(in: list<borrow<pollable>>) -> list<u32>;
  }
}


package example:ex2 {
  interface one {
    record custom {
      a: u64,
      b: u64,
      c: u64,
    }

    writer: func() -> string;

    add-from-record: func(data: custom) -> u64;
  }
}
```

here is the JS I've run through jco:

```
function writer() {
    return "Hello world!";
}
function addFromRecord({a, b, c}) {
    return a+b+c;
}

export const one = {
    writer,
    addFromRecord
}

export function run() {
    return writer();
}
```

For this WIT, the parser shows only:
```
iex(8)> Wasmex.Native.wit_exported_functions(".", File.read!("ex2.wit"))
%{"run" => 0}
```

The wasm documentation hides or fails to clearly disclose what an exported function via an interface would look like in string form to return the index using get_func. My hope was the WIT parser would give me a clue to that name. So I've tried all kinds of string paths, none yield success from get_func. For example:
```
iex(17)> Example.call(i, "example:ex2/one/writer", [])
{:error, "Function example:ex2/one/writer not exported."}
```

I don't know what the proper name for `writer` is.

I did produce a WAT and the tail section, with exports, looks like:
```
  (export (;2;) "run" (func 1))
  (type (;2;) (func (result string)))
  (alias core export 2 "example:ex2/one#writer" (core func (;5;)))
  (alias core export 2 "cabi_post_example:ex2/one#writer" (core func (;6;)))
  (func (;3;) (type 2) (canon lift (core func 5) (memory 0) string-encoding=utf8 (post-return 6)))
  (type (;3;) (record (field "a" u64) (field "b" u64) (field "c" u64)))
  (type (;4;) (func (param "data" 3) (result u64)))
  (alias core export 2 "example:ex2/one#add-from-record" (core func (;7;)))
  (alias core export 2 "cabi_post_example:ex2/one#add-from-record" (core func (;8;)))
  (func (;4;) (type 4) (canon lift (core func 7) (post-return 8)))
  (component (;0;)
    (type (;0;) (func (result string)))
    (import "import-func-writer" (func (;0;) (type 0)))
    (type (;1;) (record (field "a" u64) (field "b" u64) (field "c" u64)))
    (import "import-type-custom" (type (;2;) (eq 1)))
    (type (;3;) (func (param "data" 2) (result u64)))
    (import "import-func-add-from-record" (func (;1;) (type 3)))
    (type (;4;) (record (field "a" u64) (field "b" u64) (field "c" u64)))
    (export (;5;) "custom" (type 4))
    (type (;6;) (func (result string)))
    (export (;2;) "writer" (func 0) (func (type 6)))
    (type (;7;) (func (param "data" 5) (result u64)))
    (export (;3;) "add-from-record" (func 1) (func (type 7)))
  )
  (instance (;1;) (instantiate 0
      (with "import-func-writer" (func 3))
      (with "import-func-add-from-record" (func 4))
      (with "import-type-custom" (type 3))
    )
  )
  (export (;2;) "example:ex2/one" (instance 1))
  (@producers
    (processed-by "wit-component" "0.219.1")
    (processed-by "ComponentizeJS" "0.14.0")
    (language "JavaScript" "")
  )
)
```

**Comment by superchris** [comment]
First of all, thanks for trying out the component support and giving us some feedback! I think you are hitting a few things we don't support just yet but I should be able to tweak what you have slightly to work given our current limitations. Documentation on component support is also just not there yet. Specifically, we do not yet support resources or exported interfaces, only function exports work just now. I'll go ahead and add additional issues for these limitations so we can track our progress. That said, I'll have a go at reworking what you have so we can reference it as an example (presuming I get it working!).

**Comment by superchris** [comment]
Hey @danj3 I was able to take your example and get it working with a few changes. I started an examples repo and put it here: https://github.com/superchris/wasm_component_example. Love to hear your thoughts, and thanks again!

**Reply by danj3** [reply]
This makes perfect sense. Using the direct function exports and this works. The types work really well:

```
  type country-list = list<tuple<string, u64>>;
  export list-countries: func() -> country-list;
```

returns exactly a list of tuples that can be sent to Map.new, even though JS has no tuples, it's really a list(list()).

I'll keep watch for developments here and look forward to using and testing more of the WIT coverage. Great work!
